articles/130_ros_time.md

```diff
@@ -39,85 +45,110 @@ Slower than real time simulation is necessary for complicated systems where accu
+- The current real time factor (also called the "multiplier", typically a value of one, but could be zero for paused or negative when going backwards).
+- The system clock time (in nanoseconds since the Unix epoch) when the message was created (or -1 if not available)
 
-To implement the time abstraction the following approach will be used.
+This message cannot be latched; to use the real time factor you will need to grab a frequency count (aka, start a stopwatch) when the message is received.
+Stale data would make this mechanism fail.
+An exception to this is when all nodes have a system clock available and the message contains the system clock stamp.
+This sytem timestamp could be used to determine the staleness of the "current time" in the message.
+Nodes that publish data before they receive the current time would also be problematic.
+It will be helpful to have the clock publisher publish a new message whenever there is a new subscription added to reduce this period of unknown time.
 
-The time abstraction can be published by one source on the `/clock` topic.
-The topic will contain the most up to date time for the ROS system.
-If a publisher exists for the topic, it will override the system time when using the ROS time abstraction.
-If `/clock` is being published, calls to the ROS time abstraction will return the latest time received from the `/clock` topic.
-If time has not been set it will return zero if nothing has been received.
-A time value of zero should be considered an error meaning that time is uninitialized.
+### ROS Time Message Sources
 
-If the time on the clock jumps backwards, a callback handler will be invoked and be required to complete before any calls to the ROS time abstraction report the new time.
-Calls that come in before that must block.
-The developer has the opportunity to register callbacks with the handler to clear any state from their system if necessary before time will be in the past.
+- An included (where?) Steady Clock Publisher.
+- An included (where?) System Clock Publisher.
+- An included (where?) Gazebo Plugin.
+- The ROSBag playback engine.
 
-The frequency of publishing the `/clock` as well as the granularity are not specified as they are application specific.
+#### Default Time Source
 
-##### No Advanced Estimating Clock
+Users can call `Node::now()` for the current time.
+With the assumption of a single Node instance per process, you can also get the current time from a static method: `Time::now()`.
+Nodes will automatically subscribe for the `/clock` message unless it is programmatically disabled. (Do we need to disable it? We can simply "not publish" to use the system clock.)
+If a publisher exists for the `/clock` topic, it will override the system time when using the ROS time abstraction.
+The `ROSTime` and the `now()` methods will fall back to `SystemTime` (and then `SteadyTime`) when an ROS Time Source is not active (aka, it has been disabled purposely or hasn't received a clock broadcast).
```

Comment:
Also from talking about this offline we thought that keeping the `use_sim_time` parameter (like in ROS 1) would be the best way to control it. We thought about using a heuristic to enable it, like "is /clock being published to", but there were so many issues we didn't see how to solve, like the fact that there would be a startup race condition and knowing how and when to swap back to non-simulated time.

I think the the parameter to control it should be used.

As to being able to programmatically avoid subscribing to it, I'd again say this can be covered by the `use_sim_time` parameter.